### PR TITLE
Integrate EDK2 PiSmmCpuDxeSmm driver change

### PR DESCRIPTION
**Pull request by kuqin12**
## Description

This change integrated the update from EDK2 where the MP information of the system will be pulled from the HOBs instead of gBS, which aligns with our existing implementation.

- [x] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [ ] Impacts security?
  - **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter
    validation improvement, ...
- [x] Breaking change?
  - **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
  - Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [ ] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [ ] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested

This change was tested on QEMU Q35 and booted to UEFI shell.

## Integration Instructions

N/A

### CODE REVIEW

**Comment by kuqin12** [comment]
This change requires an update of BASECORE from this PR: https://github.com/microsoft/mu_basecore/pull/839/files

File: MmSupervisorPkg/Core/Services/MpService/MpService.c [hunk 1]
```diff
@@ -1966,6 +1966,16 @@ InitializeMpSyncData (
           break;
         }
       }
+    } else {
```
Comment (kenlautner) [inline MmSupervisorPkg/Core/Services/MpService/MpService.c:1969]: This is a duplicate.

**Review by kenlautner — APPROVED** [review]

File: MmSupervisorPkg/Core/Relocate/Relocate.c [hunk 2]
```diff
@@ -517,6 +517,264 @@ LockMmCoreBeforeExit (
...
+    if (NumberOfProcessors >= MaxNumberOfCpus) {
+      break;
+    }
+
+    GuidHob = GetNextGuidHob (&gSmmBaseHobGuid, GET_NEXT_HOB (GuidHob));
+  }
+
+  ASSERT (NumberOfProcessors == MaxNumberOfCpus);
+  if (NumberOfProcessors != MaxNumberOfCpus) {
+    CpuDeadLoop ();
```
Comment (makubacki) [inline MmSupervisorPkg/Core/Relocate/Relocate.c:593]: I suggest using [`PanicLib`](https://github.com/microsoft/mu_basecore/blob/HEAD/MdePkg/Include/Library/PanicLib.h) in situations like this.

**Review by makubacki — APPROVED** [review]

File: MmSupervisorPkg/Core/Services/MpService/MpService.c (hunk 1 above)
Comment (kuqin12) [reply MmSupervisorPkg/Core/Services/MpService/MpService.c]: removed

File: MmSupervisorPkg/Core/Relocate/Relocate.c (hunk 2 above)
Comment (kuqin12) [reply MmSupervisorPkg/Core/Relocate/Relocate.c]: Updated.

**Comment by kuqin12** [comment]
Depending on https://github.com/microsoft/mu_tiano_plus/pull/272 to resolve the pipeline error.

**Pull request merged**